crates/bevy_feathers/src/focus.rs

```diff
@@ -0,0 +1,105 @@
+
+use bevy_app::{Plugin, PostUpdate};
+use bevy_ecs::{
+    change_detection::DetectChanges,
+    component::Component,
+    entity::Entity,
+    hierarchy::Children,
+    query::With,
+    reflect::ReflectComponent,
+    schedule::IntoScheduleConfigs,
+    system::{Commands, Query, Res},
+};
+use bevy_input_focus::{InputFocus, InputFocusVisible};
+use bevy_reflect::{prelude::ReflectDefault, Reflect};
+use bevy_ui::{BorderRadius, GlobalZIndex, Node, Outline, PositionType, UiSystems, Val};
+
+use crate::{theme::UiTheme, tokens};
+
+/// A marker component which indicates that this entity should display a visible focus outline
+/// when either it, or its ancestor, are focused. Insert this into a widget on the entity that
+/// you wisth to display a focus outline.
+#[derive(Component, Default, Clone, Reflect)]
+#[reflect(Component, Clone, Default)]
+pub struct VisibleFocusAnchor;
+
+/// A marker used to identify a visible focus outline.
+#[derive(Component, Default, Clone, Reflect)]
+#[reflect(Component, Clone, Default)]
+struct FocusOutline;
+
+fn focus_system(
+    mut commands: Commands,
+    focus: Res<InputFocus>,
+    focus_visible: Res<InputFocusVisible>,
+    theme: Res<UiTheme>,
+    q_focus_outlines: Query<Entity, With<FocusOutline>>,
+    q_focus_anchors: Query<&BorderRadius, With<VisibleFocusAnchor>>,
+    q_children: Query<&Children>,
+) {
+    if focus.is_changed() {
+        // Start by despawning all the existing focus outlines
+        for outline in q_focus_outlines.iter() {
+            commands.entity(outline).despawn();
+        }
+
+        // Walk the descendants of the current focus element.
+        if focus_visible.0
+            && let Some(focused) = focus.0
+        {
+            spawn_focus_rings(
+                &mut commands,
+                focused,
+                &q_children,
+                &q_focus_anchors,
+                theme.as_ref(),
+            );
+        }
+    }
+}
+
+fn spawn_focus_rings(
+    commands: &mut Commands,
+    parent: Entity,
+    q_children: &Query<&Children>,
+    q_anchors: &Query<&BorderRadius, With<VisibleFocusAnchor>>,
+    theme: &UiTheme,
+) {
+    if let Ok(radii) = q_anchors.get(parent) {
+        let outline = commands
+            .spawn((
+                Node {
+                    position_type: PositionType::Absolute,
+                    left: Val::Px(0.0),
+                    right: Val::Px(0.0),
+                    top: Val::Px(0.0),
+                    bottom: Val::Px(0.0),
+                    ..Default::default()
+                },
+                FocusOutline,
+                GlobalZIndex(100),
+                Outline {
+                    color: theme.color(tokens::FOCUS_RING),
+                    width: Val::Px(2.0),
+                    offset: Val::Px(2.0),
+                },
+                *radii,
+            ))
+            .id();
+        commands.entity(parent).add_child(outline);
```

Comment:
Nodes with a measure func like `Text` entities don't like having children, if a text node is given a focus outline like this it will probably collapse to zero size.